core/chains/evm/client/rpc_client.go

```diff
@@ -1134,6 +1200,49 @@ func Name(r *rpcClient) string {
 	return r.name
 }
 
+func (r *rpcClient) onNewHead(ctx context.Context, requestCh <-chan struct{}, head *evmtypes.Head) {
+	if head == nil {
+		return
+	}
+
+	r.stateMu.Lock()
+	defer r.stateMu.Unlock()
+	if !commonclient.CtxIsHeathCheckRequest(ctx) {
+		r.highestUserObservations.BlockNumber = max(r.highestUserObservations.BlockNumber, head.Number)
+		r.highestUserObservations.SetTotalDifficultyIfGt(head.TotalDifficulty)
+	}
+	select {
+	case <-requestCh: // no need to update latestChainInfo, as rpcClient already started new life cycle
+		return
+	default:
+		r.latestChainInfo.BlockNumber = head.Number
+		r.latestChainInfo.TotalDifficulty = head.TotalDifficulty
+	}
+}
+
+func (r *rpcClient) onNewFinalizedHead(ctx context.Context, requestCh <-chan struct{}, head *evmtypes.Head) {
+	if head == nil {
+		return
+	}
+	r.stateMu.Lock()
+	defer r.stateMu.Unlock()
+	if !commonclient.CtxIsHeathCheckRequest(ctx) {
```

Comment:
Why are we making this distinction? If a user makes a request and the returned finalized number is less than the observed, are we going to return the one that is lower?

Reply:
No, this logic does not affect returned value. 
We are making this distinction to ensure that only user's observation increase min finalized block requirement for other RPCs. 
Without this distinction we ran into a risk of a short life time of an active now as health checks constantly poll latest finalized block and we may observe "stairs" in the finalized state observed by an RPC at the moment of user's call:
1. RPC A (finalized block is 10), RPC B (finalized block is 11). B is only healthy RPC, so we are forced to cancel subscriptions and switch to B, if A was active.
2. On block 12, due to race between RPCs state and our observation, we see that RPC A is at block 12, but RPC B is still at block 11. So we are forced to cancel requests and subscriptions to B and make A - new active RPC.
3. A (12), B (13).....

With the distinction we only take into account results of an active RPC that were observed by a client.